logger/logger.go

```diff
@@ -1,120 +1,53 @@
 package logger
 
 import (
-	"fmt"
 	"io"
-	"io/ioutil"
-	"log"
-	"os"
-	"path/filepath"
-	"runtime"
 	"time"
 )
 
-func init() {
-	setDefaultOutputs()
-}
-
-// SetupLogger configure logger instance with user provided settings
-func SetupLogger() (err error) {
-	if *Logger.Enabled {
-		err = setupOutputs()
-		if err != nil {
-			return
-		}
-		logLevel()
-		if Logger.ColourOutput {
-			colourOutput()
-		}
-	} else {
-		clearAllLoggers()
+func newLogger(c *Config) *Logger {
+	return &Logger{
+		Timestamp:   c.AdvancedSettings.TimeStampFormat,
+		Spacer:      c.AdvancedSettings.Spacer,
+		ErrorHeader: c.AdvancedSettings.Headers.Error,
+		InfoHeader:  c.AdvancedSettings.Headers.Info,
+		WarnHeader:  c.AdvancedSettings.Headers.Warn,
+		DebugHeader: c.AdvancedSettings.Headers.Debug,
 	}
-	return
 }
 
-// setDefaultOutputs() this setups defaults used by the logger
-// This allows it to be used without any user configuration
-func setDefaultOutputs() {
-	debugLogger = log.New(os.Stdout,
-		"[DEBUG]: ",
-		log.Ldate|log.Ltime)
-
-	infoLogger = log.New(os.Stdout,
-		"[INFO]:  ",
-		log.Ldate|log.Ltime)
-
-	warnLogger = log.New(os.Stdout,
-		"[WARN]:  ",
-		log.Ldate|log.Ltime)
-
-	errorLogger = log.New(os.Stdout,
-		"[ERROR]: ",
-		log.Ldate|log.Ltime)
-
-	fatalLogger = log.New(os.Stdout,
-		"[FATAL]: ",
-		log.Ldate|log.Ltime)
+func SetupGlobalLogger() {
+	logger = newLogger(GlobalLogConfig)
 }
 
-// colorOutput() sets the prefix of each log type to matching colour
-// TODO: add windows support
-func colourOutput() {
-	if runtime.GOOS != "windows" || Logger.ColourOutputOverride {
-		debugLogger.SetPrefix("\033[34m[DEBUG]\033[0m: ")
-		infoLogger.SetPrefix("\033[32m[INFO]\033[0m: ")
-		warnLogger.SetPrefix("\033[33m[WARN]\033[0m: ")
-		errorLogger.SetPrefix("\033[31m[ERROR]\033[0m: ")
-		fatalLogger.SetPrefix("\033[31m[FATAL]\033[0m: ")
+func (l *Logger) newLogEvent(data, header string, w io.Writer) {
+	if w == nil {
+		return
 	}
-}
-
-// clearAllLoggers() sets all logger flags to 0 and outputs to Discard
-func clearAllLoggers() {
-	debugLogger.SetFlags(0)
-	infoLogger.SetFlags(0)
-	warnLogger.SetFlags(0)
-	errorLogger.SetFlags(0)
-	fatalLogger.SetFlags(0)
+	e := eventPool.Get().(*LogEvent)
+	e.output = w
+	e.data = e.data[:0]
+	e.data = append(e.data, []byte(header)...)
+	e.data = append(e.data, l.Spacer...)
+	if l.Timestamp != "" {
+		e.data = time.Now().AppendFormat(e.data, l.Timestamp)
+	}
+	e.data = append(e.data, l.Spacer...)
+	e.data = append(e.data, []byte(data)...)
 
-	debugLogger.SetOutput(ioutil.Discard)
-	infoLogger.SetOutput(ioutil.Discard)
-	warnLogger.SetOutput(ioutil.Discard)
-	errorLogger.SetOutput(ioutil.Discard)
-	fatalLogger.SetOutput(ioutil.Discard)
+	e.output.Write(e.data)
```

Comment:
I've noticed you append "\n" on almost every log. Is there a reason for it being sometimes added? Rather than put the onus on the individual log to have a new line (and ruin the output if they forget), I think it should be added once here (or somewhere where its only appended once). You could always check here if its already on the end before appending to prevent an additional "\n". I think the benefits of adding it here it are better than the chance that someone adds an extra empty line

Reply:
This was something we discussed on when a new line should be added

In the std logger (which is what we currently use) a new line is appended if one is not found 
```go
	if len(s) == 0 || s[len(s)-1] != '\n' {
		l.buf = append(l.buf, '\n')
	}
```
It was decided that Info/Info (and all other versions of it Debug/Debugf etc) will not append a new line and output exactly as the user inputs it while Infoln would append a new line

I went through and updated all usage of Info/Infof to append a new line as required
I am happy to revert this change if we would rather just append a new line 

@thrasher-

Reply:
Thanks for the explanation. I appreciate that its being explicit there with `log` vs` logln`. Though I also feel that the use-case for wanting to have extra data/logs on the same line from multiple `log.Debug` could easily be handled via `log.Debugf("%v%v%v")`. The downside though is if a user forgets the  "\n" in any future PR we will be asking  "are you intending to have multiple logs on the same line?". But if its ever a mistake, its something that a user will only forget once (hopefully).
So I'm okay if it stays, but think that it will be something we might run into a few times mistakenly